### DISCUSSION TITLE
2.4 Release

### DISCUSSION THREAD

**Discussion by matty0ung** [Brewtarget/brewtarget#545] [General]
Do we have a checklist of things we need to do to get the 2.4 release out the door?  I'd be happy to help get this done, but I'm not sure what specific things to tackle.

**Comment by mattiasmaahl** [comment]
To my knowledge there is no checklist so to speak, but there is that project, [Version 2.4](https://github.com/Brewtarget/brewtarget/projects/1), for the 2.4 release. 
Maybe we can clean that up and do the release that way?
I happy to help now that I'm back in the game again.

**Reply by matty0ung** [reply]
It certainly feels like we have some of the bits:

- https://github.com/Brewtarget/brewtarget/projects/1 is a list of bugs to fix for 2.4 release
- https://github.com/Brewtarget/brewtarget/wiki/Release-Checks is a list of things to check before declaring code ready to release
- https://github.com/Brewtarget/brewtarget/wiki/Mac-Packaging is how to build packages for Mac
- https://github.com/Brewtarget/brewtarget/blob/develop/dev-doc/WindowsPackaging.txt is how to build packages for Windows
- https://github.com/Brewtarget/brewtarget/wiki/Uploading-and-Releasing is more general stuff about how to do a release, but needs updating (put binaries on GitHub rather than Launchpad) and completing (how to update the brewtarget website)

I think the last one is probably the main one to work on, possibly pulling in some stuff from https://docs.github.com/en/github/administering-a-repository/managing-releases-in-a-repository and https://docs.github.com/en/github/administering-a-repository/about-releases

Glad to hear you're through the most sleepless part of being a parent! :)

**Reply by mattiasmaahl** [reply]
Yes I agree, the last one is due for an update at the least. Also Do we have someone to build Mac-packages? Or did @mikfire run on mac? I don't remember. Or do we crosscompile on Linux?
The project needs to be cleaned out of things that are done and things that will not make it into the releaase.
I'll go through that tomorrow and revert with my thoughts on that.

**Reply by xzfantom** [reply]
> 
> 
> It certainly feels like we have some of the bits:
> 
>     * https://github.com/Brewtarget/brewtarget/projects/1 is a list of bugs to fix for 2.4 release
> 
>     * https://github.com/Brewtarget/brewtarget/wiki/Release-Checks is a list of things to check before declaring code ready to release
> 
>     * https://github.com/Brewtarget/brewtarget/wiki/Mac-Packaging is how to build packages for Mac
> 
>     * https://github.com/Brewtarget/brewtarget/blob/develop/dev-doc/WindowsPackaging.txt is how to build packages for Windows
> 
>     * https://github.com/Brewtarget/brewtarget/wiki/Uploading-and-Releasing is more general stuff about how to do a release, but needs updating (put binaries on GitHub rather than Launchpad) and completing (how to update the brewtarget website)
> 

I'd add milestones, too: https://github.com/Brewtarget/brewtarget/milestone/3
Probably we should decide what system to use - projects or milestones. I'm not sure we need both. I'd vote for milestones - they are in Issues tab, so you can easily see them.
One more thing about organization of stable releases in git. Now there are stable branches and workflow is to backport changes from master to it. Looks like some waste of time, isn't adding tag to master enough? I see use of additional branch only if smbdy want to backport some patches from more recent release to older release.

**Reply by matty0ung** [reply]
Certainly if the goal is to follow https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow then releases end up as tags on master.  Or are you talking about doing something else?

**Reply by xzfantom** [reply]
Yes, I think creating release branch and merging changes back and forth is waste of time in our case. We can do all work in develop branch and feature branches created from develop and. Then if all tasks from milestone/project are done and no active bug issues we merge develop to master and put tag on it.
May be there is some profit of using dedicated stable branches, but I don't see it. Only additional work to merge between branches.

**Reply by mikfire** [reply]
I don't like git, nor its work flows. Yall are welcome to figure out how you want the branching to work and then you will have to explain it to me like I am a 5-year-old.

I would put in a small vote for milestones. Easier is better, and I will probably ignore them equally.

**Reply by matty0ung** [reply]
FWIW, I found I liked git more after I read a good explanation of how it works: https://www.sbf5.com/~cduan/technical/git/.

For the workflows, I think there's a lot of mileage in following a tried-and-tested one (rather than inventing our own one from scratch), but it doesn't necessarily need to be Git Flow or GitHub Flow.  @xzfantom, would it be fair to say your views are close to these: https://www.endoflineblog.com/gitflow-considered-harmful?

**Reply by xzfantom** [reply]
@matty0ung Yes, I think GitFlow is overcomplicated and was talking about something like GithubFlow. The only thing - it represents continuous developing and no releases. As far as we use releases we can introduce them as tags in master branch (maybe a little more work) or tag commit that is considered as ready for release in develop branch.
Anyway it's not really big deal for me, I'll accept all suitable model)

**Reply by matty0ung** [reply]
Unless anyone objects, we could try Github Flow and see how we get on with it.  First thing would be to make a wiki page that has the CliffsNotes explanation of how to use it, both day-to-day and with making and supporting releases.  I particular, a lot of the spiel I found by googling assumes you're doing CI/CD, and doesn't really cover hotfixes.  (Presumably, to do the first hotfix for a release we need to make a branch from the relevant tag?)

**Reply by xzfantom** [reply]
Maybe instead of hotfixing release is better to create new one adding 1 to patch number?

**Reply by matty0ung** [reply]
Yes, that's an alternative approach.  The downside would be that we would have to be somewhat more careful about when to commit big changes.

Let's take an example.  Say we've done a whole bunch of testing and bug-fixing to get the 2.4 release done and out the door.  Maybe a few weeks or a few months later we then want to do a 2.4.1 release with updated translations (as has been done in the past, because it takes a while to get the translations done).  If there were no other major commits since 2.4, then, as you say, we just declare the current version of master as the new 2.4.1 release and we're done.  But, if someone checked in some major change after the 2.4 release, then that's going to end up in the 2.4.1 release.  In that case, the 2.4.1 release would no longer be "just some updated translations for 2.4", and it might well be missing translations required by whatever the post-2.4 major change was.

**Reply by xzfantom** [reply]
I agree, it can be a problem. One solution can be not accept pull requests on large commits for some time after release. It can work if translation is updated in some reasonable time. So it happens only at release time and not always as in gitflow.
One more interesting way is to use feature toggle https://en.m.wikipedia.org/wiki/Feature_toggle
I never used it and can't promote it but it's pluses is to allow people use some not tested/translated stuff if they want)

**Reply by matty0ung** [reply]
Yes, have used feature flags in the past when doing web development.  They were useful where we wanted to do frequent releases and want to commit new code without enabling it in "production" until it was complete and tested.  But it did get tricky for work that involved significant refactoring or touching a lot of code.

I'm not against just doing tag releases from a single master/trunk branch (which I think are akin to Trunk-based Development, as described in this article https://www.toptal.com/software/trunk-based-development-git-flow), just as long as everyone is signed up to the consequences of it.  I guess @mikfire would know more about the translations in particular.

**Reply by mikfire** [reply]
What I remember is this. The translations were put off into their own because they take a lot of time. Even though the strings don't change much from release to release, it is still tedious work. As a rule, open source projects aren't good at the tedious work. 

The idea was we would release a new version (2.4.0) without waiting for every language to finish. Once the translations were done, we would release a minor update (2.4.1) with the updated translations.

If I were to vote on this, I might vote to keep it as is. Yes, it is slightly more work on our part. The theory is that after we released 2.4, there would be a 2.4.1, 2.4.2, ... release as we found and fixed bugs. It is difficult to convince people to upgrade from 2.4.0 to 2.4.0. By doing the hard branch instead of a tag, we can be certain that whatever mess I've made of develop this week won't impact our ability to ship bug fixes.

**Comment by mikfire** [comment]
No. I do not own a mac. I am quite confident it was not cross-compiled, but I have no idea who actually built the mac package. I was not really involved in the release process and I bungled the hand off from rocketman.

I actually have most of the build repos up to date and have recently rolled a proper set of .deb and .rpm packages. I think I can do a .exe, but I am less confident of that.

**Reply by mattiasmaahl** [reply]
Aha, I remembered wrong then. I can look into cross-compiling for Mac and see what I can figure out.
I can easily make .Exe and NSIS installer for the release for the 2.4 version of the code.

**Reply by matty0ung** [reply]
Worst case we could get the release out on Windows and Linux and add the Mac build later, either when we've got cross-compilation working or when someone with a Mac is available.

**Reply by xzfantom** [reply]
There is mac vm in Github Actions, so we don't need to cross-compile it, just add new pipeline in CI and grab ready distrib.
